### PR TITLE
Cuda constraints for matrix-free

### PR DESCRIPTION
**Pull request by Rombur**
This add supports for hanging nodes when using matrix-free with CUDA. This is based on https://github.com/kalj/dealii-cuda/blob/710a46a6c2f98c966ad1842f18e29bcfbabbe704/matrix_free_gpu/hanging_nodes.cuh from @kalj. This doesn't work if `dim` equals 2 and `fe_degree` is even but I won't have time to work on this in the foreseeable further. I would like to merge this because it is an improvement and I want to avoid future conflict (and maybe someone smarter than me (@kalj  maybe :smile:)  can find the problem).

### CODE REVIEW

File: include/deal.II/matrix_free/cuda_hanging_nodes.h [hunk 1]
```diff
@@ -0,0 +1,972 @@
...
+                          edge_neighbor.second;
+                        mask |= line_to_edge[local_line][1] |
+                                line_to_edge[local_line][2];
+
+                        bool flipped = false;
+                        if (cell->line(local_line)->vertex_index(0) ==
+                            neighbor_cell->line(local_line_neighbor)
+                              ->vertex_index(0))
+                          {
+                            // Assuming line directions matches axes directions,
```
Comment (kronbichler) [inline include/deal.II/matrix_free/cuda_hanging_nodes.h]: matches -> match

**Comment by kronbichler** [comment]
Regarding the bug with even degree: I will try to look into it at some point, at least during the review I couldn't spot any potential problems, despite looking for them. I have to compare with @kalj's code to see whether there is a difference because we have done some computations on hanging nodes with even degrees in 2d and they worked.

File: include/deal.II/matrix_free/cuda_hanging_nodes.h [hunk 2]
```diff
@@ -0,0 +1,972 @@
...
+  /**
+   * This class creates the mask used in the treatment of hanging nodes in
+   * CUDAWrappers::MatrixFree.
+   * The implementation of this class is explained in <em>Matrix-Free
+   * Finite-Element Computations On Graphics Processors With Adaptively Refined
+   * Unstructured Meshes</em> by Karl Ljungkvist, SpringSim-HPC, 2017 April
+   * 23-26.
+   */
+  template <int dim>
+  class HangingNodes
```
[the remark this answers is not in the document]
Comment (Rombur) [reply include/deal.II/matrix_free/cuda_hanging_nodes.h]: Yes I will move it

**Comment by Rombur** [comment]
I have done the changes and I have move everything into `cuda_hanging_nodes_internal`

**Pull request merged**